File: DAYDAYUP/javaCode/IO/NIO/asynchronousFileChannel/future.java
package IO.NIO.asynchronousFileChannel;

import java.nio.ByteBuffer;
import java.nio.channels.AsynchronousFileChannel;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.StandardOpenOption;
import java.util.concurrent.Future;

/**
 * @author tianmingbo
 */
public class future {
    public static void main(String[] args) throws Exception {
        Path path = Paths.get("IO/NIO/channel/das1.txt");
        //异步读取操作
        try (AsynchronousFileChannel fileChannel = AsynchronousFileChannel.open(path, StandardOpenOption.READ)) {
            ByteBuffer buffer = ByteBuffer.allocate(1024);
            long position = 0;

            while (true) {
                //表示读取操作的结果
                Future<Integer> result = fileChannel.read(buffer, position);
                //检查异步操作是否完成
                while (!result.isDone()) {
                    // 在这里可以执行其他任务，例如处理其他 I/O 操作
                    System.out.println("正在读取数据...");
                }

                int bytesRead = result.get();
                if (bytesRead <= 0) {
                    break;
                }

                position += bytesRead;
                buffer.flip();

                byte[] data = new byte[buffer.limit()];
                buffer.get(data);
                System.out.println(new String(data));

                buffer.clear();
            }
        }

    }


}

File: Language/python/aio/aio+multi/future.py
from concurrent.futures import Future

future = Future()


def callback(future):
    print("当set_result的时候，执行回调，我也可以拿到返回值：", future.result())


# 通过调用add_done_callback方法，可以将该future绑定一个回调函数
# 这里只需要传入函数名即可，future会自动传递给callback的第一个参数
# 如果这里需要多个参数的话，怎么办呢？很简单，使用偏函数即可
future.add_done_callback(callback)

# 当什么时候会触发回调函数的执行呢？
# 当future执行set_result的时候
future.set_result("return value")
"""
当set_result的时候，执行回调，我也可以拿到返回值： return value
"""
